File: Money.API/Models/Category.cs
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;

namespace Money.API.Models
{
    public class Category
    {
        public Category()
        {
            Transactions = new HashSet<Transaction>();
        }

        public int Id { get; set; }

        [Required]
        [StringLength(200)]
        public string? Name { get; set; }

        public decimal? Budget { get; set; }

        public int? ParentId { get; set; }

        [JsonIgnore]
        public virtual Category Parent{ get; set; }
        public virtual ICollection<Category> Children { get; set; }
        [JsonIgnore]
        public virtual ICollection<Transaction> Transactions { get; set; }
    }
}

File: Money/src/app/models/category.ts
/**
 * Transaction's category
 */
export interface Category {
    /**
     * Unique idenfier.
     */
    id: number;

    /**
     * Name of the category.
     */
    name: string;

    /**
     * Budget of the category
     */
    budget: number;

    /**
     * Parent category or null if it has no parent.
     */
    parentId?: number;
    
    /**
     * Children categories or null if it has no child.
     */
    children?: Category[];
}